src/models/ModelBuilder.cc

```diff
@@ -3,23 +3,136 @@
+
+    if (not hasTheoryFunction(sr)) {
+        vec<PTRef> formalArgs; formalArgs.capacity(vals.size());
+        std::string symName(logic.getSymName(sr));
+        // Ensure that no formal arg name collides with the function name
+        std::string formalArgPrefix(Model::getFormalArgBaseNameForSymbol(logic, sr, formalArgDefaultPrefix));
+
+        for (PTRef v : vals) {
+            std::stringstream ss;
+            ss << formalArgPrefix << uniqueNum++;
+            formalArgs.push(logic.mkVar(logic.getSortRef(v), ss.str().c_str()));
+        }
+        FunctionSignature templateSig(logic.protectName(logic.getSymName(sr)), std::move(formalArgs), logic.getSortRef(sr));
+        definitions.insert({sr,opensmt::pair<FunctionSignature,ValuationNode*>{std::move(templateSig), nullptr}});
+    }
+    auto & signatureAndValuation = definitions.at(sr);
+    FunctionSignature & templateSignature = signatureAndValuation.first;
+    const vec<PTRef> & formalArgs = templateSignature.getArgs();
+    vec<opensmt::pair<PTRef,PTRef>> valuation; valuation.capacity(vals.size());
+    for (int i = 0; i < vals.size(); i++) {
+        valuation.push({formalArgs[i], vals[i]});
+    }
+    assert(logic.getSortRef(val) == logic.getSortRef(sr));
+    signatureAndValuation.second = addToValuationTree(valuation, val, signatureAndValuation.second);
+}
+
+ModelBuilder::ValuationNode * ModelBuilder::addToValuationTree(const vec<opensmt::pair<PTRef,PTRef>> & valuation, PTRef value, ValuationNode * root) {
+    assert(valuation.size() >= 1);
+
+    if (root == nullptr) {
+        root = valuationNodeFactory.getValuationNode(valuation[0].first, PTRef_Undef);
+    }
+
+    ValuationNode * current = root;
+
+    for (int i = 0; i < valuation.size(); i++) {
+        auto & varVal = valuation[i];
+        PTRef val = varVal.second; // The value of my parent's var
+
+        assert(current != nullptr);
+
+        bool foundChild = false;
+        for (auto child: *current) {
+            if (child->getValue() == val) {
+                current = child;
+                foundChild = true;
+                break;
+            }
+        }
+        if (foundChild) {
             continue;
+        } else {
+            // No child found with this value.  Need to create a new child.  If this will be a leaf
+            // (i.e., it is the last value in a valuation) the var will by convention be set to the
+            // value of this valuation
+            current = current->addChild(valuationNodeFactory.getValuationNode(i == valuation.size()-1 ? value : valuation[i+1].first, val));
         }
-        assert(logic.isVar(key));
-        PTRef target = copy[key];
-        PTRef val = model->evaluate(target);
-        assert(logic.isConstant(val));
-        assignCopy.insert(std::make_pair(key, val));
     }
-    this->assignment = std::move(assignCopy);
+    return root;
 }
+
+// Copied from TermVisitor.  Maybe it'd be possible to template TermVisitor to handle also ValuationNodes
```

Comment:
Maybe we should try to build some kind of generic DAG visitor...

Reply:
Indeed, something like that.